tests/callbacks_test.cpp

```diff
@@ -0,0 +1,462 @@
+                       MatType& /* coordinates */)
+  { calledEndOptimization = true; }
+
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EvaluateConstraint(OptimizerType& /* optimizer */,
+                          FunctionType& /* function */,
+                          const MatType& /* coordinates */,
+                          const size_t /* constraint */,
+                          const double /* constraintValue */)
+  { calledEvaluateConstraint = true; }
+
+  template<typename OptimizerType,
+           typename FunctionType,
+           typename MatType,
+           typename GradType>
+  void GradientConstraint(OptimizerType& /* optimizer */,
+                          FunctionType& /* function */,
+                          const MatType& /* coordinates */,
+                          const size_t /* constraint */,
+                          GradType& /* gradient */)
+  { calledGradientConstraint = true; }
+
+  bool calledEvaluate;
+  bool calledGradient;
+  bool calledBeginEpoch;
+  bool calledEndEpoch;
+  bool calledBeginOptimization;
+  bool calledEndOptimization;
+  bool calledEvaluateConstraint;
+  bool calledGradientConstraint;
+};
+
+template<typename OptimizerType>
+void CallbacksFullFunctionTest(OptimizerType& optimizer,
+                               bool calledEvaluate,
+                               bool calledGradient,
+                               bool calledBeginEpoch,
+                               bool calledEndEpoch,
+                               bool calledBeginOptimization,
+                               bool calledEndOptimization,
+                               bool calledEvaluateConstraint,
+                               bool calledGradientConstraint)
+{
+  arma::mat data, testData, shuffledData;
+  arma::Row<size_t> responses, testResponses, shuffledResponses;
+
+  LogisticRegressionTestData(data, testData, shuffledData,
+      responses, testResponses, shuffledResponses);
+  LogisticRegression<> lr(shuffledData, shuffledResponses, 0.5);
+
+  CompleteCallbackTestFunction cb;
+
+  arma::mat coordinates = lr.GetInitialPoint();
+  optimizer.Optimize(lr, coordinates, cb);
+
+  REQUIRE(cb.calledEvaluate == calledEvaluate);
+  REQUIRE(cb.calledGradient == calledGradient);
+  REQUIRE(cb.calledBeginEpoch == calledBeginEpoch);
+  REQUIRE(cb.calledEndEpoch == calledEndEpoch);
+  REQUIRE(cb.calledBeginOptimization == calledBeginOptimization);
+  REQUIRE(cb.calledEndOptimization == calledEndOptimization);
+  REQUIRE(cb.calledEvaluateConstraint == calledEvaluateConstraint);
+  REQUIRE(cb.calledGradientConstraint == calledGradientConstraint);
+}
+
+/**
+ * Make sure we invoke all callbacks (AdaDelta).
+ */
+TEST_CASE("AdaDeltaCallbacksFullFunctionTest", "[CallbacksTest]")
+{
+  AdaDelta optimizer(1.0, 1, 0.99, 1e-8, 3, 1e-9, true);
+  CallbacksFullFunctionTest(optimizer, true, true, true, true, true, true,
+      false, false);
+}
```

Comment:
Nice, this is a good way to test that all callbacks are called, and it's easy to add this test for new optimizers.